### PR TITLE
raftstore: reduce the overhead of region heartbeat

### PR DESCRIPTION
**Pull request by Connor1996**
###  What have you changed?
This PR greatly reduces the overhead of region heartbeat
- Using vec to replace btreemap for `RangeProperties`
- Updating region size/keys if it is not calculated ever when heartbeating
- Updating region size/keys with accurate value when full scanning

Here is the performance difference: reduce almost 50% overhead
before:
屏幕快照 2019-09-26 上午9 35 11
after:
屏幕快照 2019-09-26 上午9 35 01

And bench result:
before:
```
tests::bench_get_approximate_size ... bench:      40,180 ns/iter (+/- 3,209)
```
after:
```
tests::bench_get_approximate_size ... bench:      29,150 ns/iter (+/- 1,690)
```

###  What is the type of the changes?
- Enhancement

###  How is the PR tested?
origin test

###  Does this PR affect documentation (docs) or should it be mentioned in the release notes?
No

###  Does this PR affect `tidb-ansible`?
No

### CODE REVIEW

File: src/raftstore/store/worker/pd.rs [hunk 1]
```diff
@@ -818,10 +818,40 @@ impl<T: PdClient> Runnable<Task> for Runner<T> {
...
-                    get_region_approximate_keys(&self.db, &region).unwrap_or_default()
+                    if let Ok(keys) = get_region_approximate_keys(&self.db, &region) {
+                        // send it to raftstore to update region approximate size
+                        let res = CasualMessage::RegionApproximateKeys { keys };
+                        if let Err(e) = self
+                            .router
+                            .send(region.get_id(), PeerMsg::CasualMessage(res))
+                        {
+                            warn!(
+                                "failed to send approximate region size";
```
Comment (Hoverbear) [inline src/raftstore/store/worker/pd.rs]:
```suggestion
                                "failed to send approximate region keys";
```

File: src/raftstore/coprocessor/properties.rs [hunk 2]
```diff
@@ -570,12 +589,46 @@ impl RangeProperties {
...
+                if idx == 0 {
+                    return vec![];
+                } else {
+                    idx - 1
+                }
+            }
+            Err(next_idx) => next_idx - 1,
+        };
+
+        self.offsets.drain(start_offset..=end_offset).collect()
```
Comment (zhangjinpeng1987) [inline src/raftstore/coprocessor/properties.rs]: seems end_offset maybe less than start_offset.

File: src/raftstore/coprocessor/properties.rs [hunk 3]
```diff
@@ -570,12 +589,46 @@ impl RangeProperties {
         end_offset - start_offset
     }
 
+    // equivalent to range(Excluded(start_key), Excluded(end_key))
+    pub fn take_excluded_range(
```
Comment (zhangjinpeng1987) [inline src/raftstore/coprocessor/properties.rs]: Please add test for this function.

File: src/raftstore/coprocessor/properties.rs [hunk 4]
```diff
@@ -620,6 +620,10 @@ impl RangeProperties {
             Err(next_idx) => next_idx - 1,
         };
 
+        if start_offset > end_offset {
```
Comment (zhangjinpeng1987) [inline src/raftstore/coprocessor/properties.rs]: s/>/>=

File: src/raftstore/coprocessor/properties.rs [hunk 5]
```diff
@@ -1056,6 +1112,8 @@ mod tests {
                 keys
             );
         }
+
+        assert!(props.take_excluded_range(b"c", b"c").is_empty());
```
Comment (zhangjinpeng1987) [inline src/raftstore/coprocessor/properties.rs]: Please add more cases for take_excluded_range.

File: src/raftstore/coprocessor/properties.rs (hunk 4 above)
Comment (Connor1996) [reply src/raftstore/coprocessor/properties.rs]: when start_offset == end_offset, the result of `drain(start_offset..=end_offset)` is not empty

File: tests/integrations/raftstore/test_split_region.rs [hunk 6]
```diff
@@ -863,3 +863,62 @@ fn test_split_with_epoch_not_match() {
...
+    // let update approximate size and keys in pd_worker first, then in split_checker
+    cluster.cfg.raft_store.split_region_check_tick_interval = ReadableDuration::millis(500);
+    cluster.cfg.raft_store.pd_heartbeat_tick_interval = ReadableDuration::millis(100);
+    cluster.start().unwrap();
+
+    let left = pd_client.get_region(b"").unwrap();
+    let right = pd_client.get_region(&max_key).unwrap();
+    assert_eq!(left, right);
+
+    thread::sleep(Duration::from_millis(800));
```
Comment (zhangjinpeng1987) [inline tests/integrations/raftstore/test_split_region.rs]: Sleep is not recommended in test, it will make test takes more time.

**Pull request merged**